### PR TITLE
[spirv] Support Hull shader translation

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

File: tools/clang/test/CodeGenSPIRV/hull.pcf.output-patch.hlsl [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+// CHECK: OpEntryPoint TessellationControl %main "main" {{%\w+}} {{%\w+}} {{%\w+}} %out_var_hullEntryPointOutput {{%\w+}}
+
+// CHECK:               %_arr_BEZIER_CONTROL_POINT_uint_16 = OpTypeArray %BEZIER_CONTROL_POINT %uint_16
+// CHECK:   %_ptr_Output__arr_BEZIER_CONTROL_POINT_uint_16 = OpTypePointer Output %_arr_BEZIER_CONTROL_POINT_uint_16
+// CHECK: %_ptr_Function__arr_BEZIER_CONTROL_POINT_uint_16 = OpTypePointer Function %_arr_BEZIER_CONTROL_POINT_uint_16
+// CHECK:                                   [[fType:%\d+]] = OpTypeFunction %HS_CONSTANT_DATA_OUTPUT %_ptr_Function__arr_BEZIER_CONTROL_POINT_uint_16
+
+// CHECK: %out_var_hullEntryPointOutput = OpVariable %_ptr_Output__arr_BEZIER_CONTROL_POINT_uint_16 Output
+// CHECK:    %temp_hullEntryPointOutput = OpVariable %_ptr_Function__arr_BEZIER_CONTROL_POINT_uint_16 Function
```
Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/hull.pcf.output-patch.hlsl]: I'd suggest to add `// CHECK-LABEL: %main = OpFunction` or some comments to make it clear the following is in the wrapper.

File: tools/clang/test/CodeGenSPIRV/hull.pcf.input-patch.hlsl [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+#include "bezier_common_hull.hlsl"
+
+// Test: PCF takes the input control points (InputPatch)
+
+// CHECK: OpEntryPoint TessellationControl %main "main" %in_var_hullEntryPointInput {{%\w+}}
+
+// CHECK:              [[fType:%\d+]] = OpTypeFunction %HS_CONSTANT_DATA_OUTPUT %_ptr_Function__arr_VS_CONTROL_POINT_OUTPUT_uint_16
+// CHECK: %in_var_hullEntryPointInput = OpVariable %_ptr_Input__arr_VS_CONTROL_POINT_OUTPUT_uint_16 Input
+// CHECK:               %param_var_ip = OpVariable %_ptr_Function__arr_VS_CONTROL_POINT_OUTPUT_uint_16 Function
```
Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/hull.pcf.input-patch.hlsl]: I'd suggest to add `// CHECK-LABEL: %main = OpFunction` or some comments to make it clear the following is in the wrapper.

File: tools/clang/test/CodeGenSPIRV/hull.pcf.primitive-id-2.hlsl [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+// Note that in this test, the main entry point *DOES NOT* take the PrimitiveID as input.
+
+
+// CHECK: OpEntryPoint TessellationControl %main "main" {{%\w+}} {{%\w+}} {{%\w+}} %gl_PrimitiveID {{%\w+}}
+
+// CHECK: OpDecorate %gl_PrimitiveID BuiltIn PrimitiveId
+
+// CHECK:     [[fType:%\d+]] = OpTypeFunction %HS_CONSTANT_DATA_OUTPUT %_ptr_Function_uint
+// CHECK:    %gl_PrimitiveID = OpVariable %_ptr_Input_uint Input
+// CHECK: %param_var_PatchID = OpVariable %_ptr_Function_uint Function
```
Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/hull.pcf.primitive-id-2.hlsl]: I'd suggest to add `// CHECK-LABEL: %main = OpFunction` or some comments to make it clear the following is in the wrapper.

File: tools/clang/test/CodeGenSPIRV/hull.pcf.primitive-id.hlsl [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+// Note that in this test, the main entry point has also taken the PrimitiveID as input.
+
+
+// CHECK: OpEntryPoint TessellationControl %main "main" {{%\w+}} {{%\w+}} %gl_PrimitiveID {{%\w+}}
+
+// CHECK: OpDecorate %gl_PrimitiveID BuiltIn PrimitiveId
+
+// CHECK:     [[fType:%\d+]] = OpTypeFunction %HS_CONSTANT_DATA_OUTPUT %_ptr_Function_uint
+// CHECK:    %gl_PrimitiveID = OpVariable %_ptr_Input_uint Input
+// CHECK: %param_var_PatchID = OpVariable %_ptr_Function_uint Function
```
Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/hull.pcf.primitive-id.hlsl]: I'd suggest to add `// CHECK-LABEL: %main = OpFunction` or some comments to make it clear the following is in the wrapper.

File: tools/clang/test/CodeGenSPIRV/hull.pcf.void.hlsl [hunk 5]
```diff
@@ -0,0 +1,32 @@
+// Run: %dxc -T hs_6_0 -E main
+
+#include "bezier_common_hull.hlsl"
+
+// CHECK: [[fType:%\d+]] = OpTypeFunction %HS_CONSTANT_DATA_OUTPUT
+// CHECK:       {{%\d+}} = OpFunctionCall %HS_CONSTANT_DATA_OUTPUT %PCF
```
Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/hull.pcf.void.hlsl]: I'd suggest to add `// CHECK-LABEL: %main = OpFunction` or some comments to make it clear the following is in the wrapper.

File: tools/clang/test/CodeGenSPIRV/hull.structure.hlsl [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+// This test ensures:
+// 1- The hull shader entry point function is invoked.
+// 2- Control barrier is placed before PCF is called.
+// 3- PCF is called by invocationID of 0.
+
+// CHECK:         %main = OpFunction %void None {{%\d+}}
+// CHECK:   [[id:%\d+]] = OpLoad %uint %gl_InvocationID
+
+// CHECK:      {{%\d+}} = OpFunctionCall %BEZIER_CONTROL_POINT %src_main %param_var_ip %param_var_i %param_var_PatchID
+
```
Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/hull.structure.hlsl]: Need to check that we've stored the result before the barrier.

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 7]
```diff
@@ -4815,10 +4909,24 @@ bool SPIRVEmitter::emitEntryFunctionWrapper(const FunctionDecl *decl,
     // initialize the corresponding temporary variable
     if (!param->getAttr<HLSLOutAttr>()) {
       uint32_t loadedValue = 0;
-      if (!declIdMapper.createStageInputVar(param, &loadedValue))
+      if (TypeTranslator::isInputPatch(param->getType())) {
+        const uint32_t hullInputPatchId =
+            declIdMapper.createStageVarWithoutSemantics(
```
[the remark this answers is not in the document]
Comment (antiagainst) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Discussed offline. It's fine to keep like this.

**Comment by antiagainst** [comment]
Also need to update the doc. But that can happen in another pull request.

File: tools/clang/test/CodeGenSPIRV/hull.structure.hlsl (hunk 6 above)
Comment (ehsannas) [reply tools/clang/test/CodeGenSPIRV/hull.structure.hlsl]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 8]
```diff
@@ -4791,13 +4791,78 @@ void SPIRVEmitter::AddExecutionModeForEntryPoint(uint32_t entryPointId) {
...
+            .Case("triangle_cw", ExecutionMode::VertexOrderCw)
+            .Case("triangle_ccw", ExecutionMode::VertexOrderCcw)
+            .Default(ExecutionMode::Max);
+    // TODO: There is no SPIR-V equivalent for "line" topology. Is it the
+    // default?
+    if (topology != "line") {
+      if (hsExecMode != spv::ExecutionMode::Max) {
+        theBuilder.addExecutionMode(entryFunctionId, hsExecMode, {});
+      } else {
+        emitError("unknown partitioning scheme in hull shader",
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Copy-pasta! :)

File: tools/clang/lib/SPIRV/SPIRVEmitter.h [hunk 9]
```diff
@@ -370,6 +370,14 @@ class SPIRVEmitter : public ASTConsumer {
   /// shader model.
   void AddExecutionModeForEntryPoint(uint32_t entryPointId);
 
+  /// \brief Adds necessary execution modes for the hull shader based on the
+  /// HLSL attributes of the entry point function.
+  /// Also, if the 'outputcontrolpoints' attribute is present, the number of
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.h]: Also writes the number of output control points to *numOutputControlPoints.

**Pull request merged**